python/prairielearn.py

```diff
@@ -52,6 +53,78 @@ class QuestionData(TypedDict):
 class ElementTestData(QuestionData):
     test_type: Literal['correct', 'incorrect', 'invalid']
 
+def grade_question_parameterized(data: QuestionData,
+                                 question_name: str,
+                                 grade_function: Callable[[Any], Tuple[Union[bool, float], Optional[str]]],
+                                 weight: int = 1) -> None:
+    '''
+    Grade question question_name, marked correct if grade_function(student_answer) returns True in
+    its first argument. grade_function should take in a single parameter (which will be the submitted
+    answer) and return a 2-tuple.
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    '''
+
+    # Create the data dictionary at first
+    data['partial_scores'][question_name] = {
+            'score': 0.0,
+            'weight': weight
+    }
+
+    try:
+        submitted_answer = data['submitted_answers'][question_name]
+    except KeyError:
+        # Catch error if no answer submitted
+        data["format_errors"][question_name] = 'No answer was submitted'
+        return
+
+    # Try to grade, exiting if there's an exception
+    try:
+        result, feedback_content = grade_function(submitted_answer)
+
+        # Check _must_ be done in this order. Int check is to deal with subclass issues
+        if isinstance(result, bool):
+            partial_score = 1.0 if result else 0.0
+        elif isinstance(result, (float, int)):
+            assert 0.0 <= result <= 1.0
+            partial_score = result
+        else:
+            assert_never(result)
+
+    except ValueError as err:
+        # Exit if there's a format error
+        data["format_errors"][question_name] = str(err)
+        return
+
+
+    # Set question score if grading succeeded
+    data['partial_scores'][question_name]['score'] = partial_score
+
+    # Put all feedback here
+    if feedback_content:
+        data['partial_scores'][question_name]['feedback'] = feedback_content
+
+def determine_score_params(score: Optional[float]) -> Tuple[str, float]:
```

Comment:
Why do we want to allow score to be `None`?

Reply:
This is used to read the score directly from the data dict, which may be None. Not allowing the none here means we just have to check for it at the call site for this function anyway.